File: doubly_linked_list/doublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
	int data;
	struct node *next;
	struct node *prev;
};

struct node *head = NULL, *last = NULL;

struct node *createNode(int key)
{
	head = (struct node *)malloc(sizeof(struct node)); //allocate the required space to *node variable
	head->data = key;
	head->next = NULL;
	head->prev = NULL;
	last = head;
	return head;
}

struct node *additionOfNodeAtBeginning(int key) //adding at the beginning of the linked list
{
	struct node *temp;
	temp = (struct node *)malloc(sizeof(struct node));
	temp->data = key;
	temp->next = head; //making temp variable point to head of the linked list
	temp->prev = NULL; //pointing the prev of temp to NULL pointer
	head->prev = temp;
	head = temp;
	return head;
}

struct node *additionOfNodeAtEnding(int key) //adding the element at the last of the linked list
{
	struct node *temp;
	temp = (struct node *)malloc(sizeof(struct node));
	temp->data = key;
	temp->next = NULL;
	temp->prev = last;
	last->next = temp;
	temp->next = NULL;
	last = temp;
	return last;
}

struct node *additionOfNodeAtRandomeIndex(int key, int index) //adding the element at the given index
{
	struct node *temp;
	temp = (struct node *)malloc(sizeof(struct node));
	temp->data = key;
	temp->next = NULL;
	if (index == 0)
	{
		return additionOfNodeAtBeginning(key);
	}
	struct node *temp1, *temp2;
	temp1 = head;
	int i = 0;
	//iterating through the list till the iterator variable is index or list is empty
	while (i < index - 1 && temp1->next != NULL)
	{
		temp1 = temp1->next;
		i++;
	}
	if (temp1->next == NULL)
	{
		additionOfNodeAtEnding(key);
	}
	temp2 = temp1->next;
	temp->prev = temp1;
	temp1->next = temp;
	temp->next = temp2;
	temp2->prev = temp;
	return head;
}

struct node *deletingFromBeginning() //deleting the element at head
{
	struct node *temp;
	temp = head;
	head = head->next; //making the head pointer to point the element next to it
	head->prev = NULL;
	free(temp); //free the space occupied by former head pointer
	return head;
}

struct node *deletingFromEnding() //deleting the element from the ending
{
	struct node *temp, *temp2;
	temp = last->prev; //temp pointer pointing to the second last element of linked list
	temp2 = last;	   //temp2 pointing to last element
	temp->next = NULL;
	last = temp; //make second last pointer as last element
	free(temp2);
	return head;
}

struct node *deletingFromRandomIndex(int index) //deleting the element from given index
{
	struct node *temp, *temp2;
	temp = head;
	int i = 0;
	while (temp != NULL && i < index)
	{
		i++;
		temp = temp->next;
	}
	if (index == 0)
	{
		return deletingFromBeginning();
	}
	else if (temp == last)
	{
		return deletingFromEnding();
	}
	else if (temp != NULL) //if temp is NULL the index is out of Bound
	{
		temp->prev->next = temp->next;
		temp->next->prev = temp->prev;
		free(temp);
	}
	else
	{
		printf("Index Out Of Bound\n"); //index is out of bound
	}
	return head;
}

void displayInReverse() //display the linked list in reverse order
{
	printf("the Linked List in Reverse Order is ");
	struct node *temp = last; //creating iterator pointing to the last element
	while (temp != NULL)
	{
		printf("%d ", temp->data);
		temp = temp->prev;
	}
	printf("\n\n");
}

void display() //display the linked list
{
	printf("the Linked List is ");
	struct node *temp = head; //creating iterator pointing to the head
	while (temp != NULL)
	{
		printf("%d ", temp->data);
		temp = temp->next;
	}
	printf("\n\n");
}

int main()
{
	struct node *root = NULL;
	root = createNode(4); // creating the list

	additionOfNodeAtBeginning(6); //adding the value 6 at beginning
	additionOfNodeAtEnding(90);	  //adding value 90 at last
	printf("After adding 90 at last ,");
	display(); //display the list
	//index are starting from 0
	additionOfNodeAtRandomeIndex(12, 1); //adding 12 at index 1
	printf("After adding 12 at index 1 ,");
	display();
	additionOfNodeAtBeginning(1);
	additionOfNodeAtEnding(2);
	additionOfNodeAtBeginning(5);
	additionOfNodeAtEnding(98);
	additionOfNodeAtRandomeIndex(122, 2); //index are starting from 0
	display();
	deletingFromBeginning(); //deleting the head element
	printf("After Deletion from beginning, ");
	display();
	deletingFromEnding(); //deleting the last element from the list
	printf("After deleting from end, ");
	display();
	deletingFromRandomIndex(3); //deleting the element at index 3(0 is starting index)
	printf("After deleting from index 3 ,");
	display();

	displayInReverse(); //display the linked list in reverse order
}

File: doubly_linked_list/doublyLinkedList.java
public class DoublyLinkedList {

    protected Node firstNode;
    protected Node lastNode;
    protected int size;

    public DoublyLinkedList(){this.size=0;}

    public void getLastNode(){
        Node temp = firstNode;
        while(temp.next != null){
            temp = temp.next;
        }
        lastNode = temp;
    }
    
    public void insertFirst(Object val){
        Node node = new Node(val);
        node.next = firstNode;
        node.prev = null;
        if (firstNode != null) {
            firstNode.prev = node;
        }
        firstNode = node;
        size ++;
    }

    public void insertLast(Object val){
        getLastNode();
        Node newNode = new Node(val);
        lastNode.next = newNode;
        newNode.prev = lastNode;
        size++;
    }

    public void checkpos(int pos){
        if(pos<0 || pos>size){
            System.out.println("Wrong Position Entered");
            System.exit(0);
        }
    }

    public void checkposition(int pos){
        if(pos<0 || pos>=size){
            System.out.println("Wrong Position Entered");
            System.exit(0);
        }
    }

    public void add(int pos, Object val){
        checkpos(pos);
        Node newNode = new Node(val);
        Node temp = firstNode;
        if(pos == 0){
            insertFirst(val);
        }else {
            while (pos-- > 1) {
                temp = temp.next;
            }
            newNode.next = temp.next;
            if(temp.next !=null){
                temp.next.prev = newNode;}
            newNode.prev = temp;
            temp.next = newNode;
            size++;
        }
    }

    public int size(){
        return size;
    }

    public void display(){
        Node temp = firstNode;
        System.out.print("START ");
        while(temp != null){
            System.out.print(temp.data+" <-> ");
            temp = temp.next;
        }
        System.out.print("END\n");
    }

    public Object get(int pos){
        checkposition(pos);
        Node temp = firstNode;
        while(pos-- >0){
            temp = temp.next;
        }
        return temp.data;
    }

    public Object remove(int pos){
        checkposition(pos);
        Node temp = firstNode;
        Node RE = new Node();
        if(pos == 0){
            RE = firstNode;
            firstNode = temp.next;
            temp.next.prev = null;
            size--;
            return RE.data;
        }
        else if(pos == size - 1){
            getLastNode();
            RE = lastNode;
            temp = lastNode;
            lastNode = temp.prev;
            temp.prev.next = null;
            size--;
            return RE.data;
        }
        while(pos-- > 1){
            temp = temp.next;
        }
        RE = temp.next;
        temp.next = temp.next.next;
        temp.next.prev = temp;
        size--;
        return RE.data;
    }

    public void reverseDisplay(){
        getLastNode();
        Node last = lastNode;
        System.out.print("START ");
        while(last != null){
            System.out.print(last.data+" <-> ");
            last = last.prev;
        }
        System.out.print("END\n");
    }

    public static void main(String[] args) {
        DoublyLinkedList d = new DoublyLinkedList();
        d.insertFirst(21);
        d.insertFirst(23);
        d.insertFirst(24);
        d.add(3,"hello");
        d.display();
        System.out.println("Size of the list is: "+ d.size());
        System.out.println("Removing Element");
        System.out.println(d.remove(3)); //displays the removed element
        d.display(); //displays the list after removal
        d.add(1,"hello"); //adds hello at 1st index
        d.insertLast(54);
        d.insertLast(421);
        System.out.println("After adding few more elements");
        d.display();
        System.out.println("Size of the list is: "+ d.size());
        System.out.println("Get Element at 0"+d.get(0));
        System.out.println("Reverse the List");
        d.reverseDisplay();
    }
}

class Node {
    Object data;
    Node next;
    Node prev;

    public Node() {}

    public Node(Object value) {
        this.data = value;
    }

    public Node(Object value, Node next, Node prev) {
        this.data = value;
        this.next = next;
        this.prev = prev;
    }
}
